R/data.table.R

```diff
@@ -1881,17 +1881,64 @@ chmatch2 <- function(x, table, nomatch=NA_integer_) {
 #    x
 #}
 
-
-as.matrix.data.table <- function(x,...)
-{
-  dm <- dim(x)
-  cn <- names(x)
+as.matrix.data.table <- function(x, rownames, ...) {
+  rn <- NULL
+  rnc <- NULL
+  if (!missing(rownames)) { # Convert rownames to a column index if possible
+    if (is.null(rownames)) {
+      warning("rownames is NULL, ignoring rownames")
+    } else if (length(rownames) == nrow(x)) {
+      # rownames argument is a vector of row names, no column in x to drop.
+      rn <- rownames
+      rnc <- NULL 
+    } else if (length(rownames) != 1) {
+      stop("rownames must be a single column in x or a vector of row names of length nrow(x)")
+    } else if (is.na(rownames)) {
+      warning("rownames is NA, ignoring rownames")
+    } else if (identical(rownames, FALSE)) {
```

Comment:
i find it a bit weird that `rownames = TRUE` is accepted, but `rownames = FALSE` is incorrect usage and results in this `warning`. Setting `rownames` dynamically by some condition evaluating to `TRUE` or `FALSE` seems like a natural use case.

Reply:
We can drop the warning